File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> container;
        std::unordered_map<char, char> hm = {
    {')', '('},
    {'}', '{'},
    {']', '['}
};

        // we need keep track of the things we have encountered in the past and use it in future.

        for (auto c : s) {
            // need make sure container is not empty!!!!
            if (!container.empty() && hm[c] == container.top()) {
                container.pop();
            } else {
                container.push(c);
            }
        }
        
        return container.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        ls = []
        hm = {
            '(': ')',
            '{': '}',
            '[': ']'
        }
        for i in s:
            if i in ['(', '{', '[']:
                ls.append(i)
                continue
            if len(ls) == 0:
                return False
            if i == hm[ls[-1]]:
                ls.pop(-1)
            else:
                return False
        if len(ls) == 0:
            return True
        return False
